### PR TITLE
discovery: add metrics + send updates from one goroutine only

### PR DESCRIPTION
**Pull request by simonpasquier**
The added metrics are:

* prometheus_sd_discovered_targets
* prometheus_sd_received_updates_total
* prometheus_sd_updates_delayed_total
* prometheus_sd_updates_total

cc @krasi-georgiev

### CODE REVIEW

File: discovery/manager.go [hunk 1]
```diff
@@ -110,24 +135,28 @@ type Manager struct {
...
-	// The sync channels sends the updates in map[targetSetName] where targetSetName is the job value from the scrape config.
+	// The sync channel sends the updates as a map where the key is the job value from the scrape config.
 	syncCh chan map[string][]*targetgroup.Group
 
 	// How long to wait before sending updates to the channel. The variable
 	// should only be modified in unit tests.
 	updatert time.Duration
+
+	// The trigger channel signals to the manager that new updates have been received from providers.
+	trigger chan struct{}
```
Comment (krasi-georgiev) [inline discovery/manager.go]: how about `triggerSend`

File: discovery/manager.go [hunk 2]
```diff
@@ -171,43 +200,48 @@ func (m *Manager) startProvider(ctx context.Context, p *provider) {
...
 			select {
-			case triggerUpdate <- struct{}{}:
+			case m.trigger <- struct{}{}:
 			default:
 			}
+		}
+	}
+}
+
+func (m *Manager) sendUpdates() {
```
Comment (krasi-georgiev) [inline discovery/manager.go]: how about just `sender()` ?

**Comment by simonpasquier** [comment]
@krasi-georgiev done!

**Comment by simonpasquier** [comment]
The benchmark doesn't show significant differences on resource usage before and after this PR but if master had the `prometheus_sd_updates_total` metric, we would probably have seen a decrease of the metric's value with the PR.

**Comment by krasi-georgiev** [comment]
LGTM

**Comment by krasi-georgiev** [comment]
don't forget to squash on merge as otherwise we end up with non build-able commits.

the merge button has a drop-down so you can select it by default.

**Pull request merged**